src/lazy/expanded/macro_table.rs

```diff
@@ -158,9 +170,23 @@ impl<'top> MacroRef<'top> {
 /// its validity and allowing evaluation to begin.
 #[derive(Debug, Clone)]
 pub struct MacroTable {
+    // Stores `Rc` references to the macro definitions to make cloning the table's contents cheaper.
     macros_by_address: Vec<Rc<Macro>>,
     // Maps names to an address that can be used to query the Vec above.
-    macros_by_name: FxHashMap<String, usize>,
+    macros_by_name: FxHashMap<Rc<str>, usize>,
+}
+
+thread_local! {
+    /// An instance of the Ion 1.1 system macro table is lazily instantiated once per thread,
+    /// minimizing the number of times macro compilation occurs.
+    ///
+    /// The thread-local instance holds `Rc` references to its macro names and macro definitions,
+    /// making its contents inexpensive to `clone()` and reducing the number of duplicate `String`s
+    /// being allocated over time.
+    ///
+    /// Each time the user constructs a reader, it clones the thread-local copy of the system macro
+    /// table.
+    pub static ION_1_1_SYSTEM_MACROS: MacroTable = MacroTable::construct_system_macro_table();
```

Comment:
> Ideally, construct_system_macro_table would be constant, though I understand that might not be possible make it const yet.

Right, constructing a heap-allocated collection is not yet possible in `const`.

> Why does the system macro table need to be cloned rather than borrowed?

(Bit of a journey here with several things you already know, bear with me.)

Each time a `Reader` is created or an `$ion_1_1` is encountered, the macro table needs to be initialized to contain the macros in `$ion`. The macros in `$ion` can also be re-exported at any time in an encoding directive. This means that in any stream, we'll end up building/using `$ion` _at least_ once.

Each macro represents some number of heap-allocated resources--names, values, annotations, etc.

In `ion-rust`, the active macro table is 'flat' in the sense that it doesn't store a tree of other tables to search; it just copies the other tables' macros into its own index/storage. To avoid needlessly performing deep copies, it stores `Rc` references to the already-constructed `Macro` instances.

If we had a static `OnceCell` holding a collection of `Macro`, the `OnceCell` could hand out `&'static Macro`. However, this would make it a different data type than the `MacroTable` we're using everywhere else. This means that the machinery for re-exporting `$ion` would need to be different from the machinery for re-exporting `$ion_encoding` or any user-defined module.

In order for the shared `$ion` macro table to hold `Rc` references, it needs to be tied to the current thread--this brings us to `thread_local`. (Note that `thread_local` is also lazily initialized.)

In aggregate this means that once per thread the system constructs/allocates a copy of the system macro table and then re-uses it across all `Reader` instantiations, `$ion_1_1`s, and re-exports for the rest of that thread and (importantly IMO) `$ion` is treated as just another module.

Thoughts?

Reply:
That was a bit of a journey—but it makes sense.